package com.example.i_go.feature_note.domain.use_case.user

import com.example.i_go.feature_note.data.remote.requestDTO.UserDTO
import com.example.i_go.feature_note.data.remote.responseDTO.UserResponseDTO
import com.example.i_go.feature_note.domain.repository.UserRepository
import com.example.i_go.feature_note.domain.util.Resource
import com.example.i_go.feature_note.domain.util.log
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import retrofit2.HttpException
import java.io.IOException
import javax.inject.Inject

class GetUserInfo @Inject constructor(
    private val repository: UserRepository,
) {
    operator fun invoke(doctor_id: Int): Flow<Resource<UserResponseDTO>> = flow {

        try {
            emit(Resource.Loading())
            val r = repository.getUserInfo(doctor_id)
            when(r.code()) {
                200 -> {
                    emit(Resource.Success(r.body()!!))
                }
                else -> {
                    "usecase ERROR ${r.code()}: ${r.errorBody().toString()}".log()
                }
            }

        } catch(e: HttpException) {
            emit(Resource.Error(e.localizedMessage ?: "An unexpected error occured"))
        } catch(e: IOException) {
            emit(Resource.Error("Couldn't reach server. Check your internet connection."))
        }
    }
}